plugins/mimecast/komand_mimecast/tasks/monitor_siem_logs/task.py

```diff
@@ -37,33 +44,62 @@ def run(  # pylint: disable=unused-argument # noqa: MC0001
         try:
             has_more_pages = False
             header_next_token = state.get(self.NEXT_TOKEN, "")
-            normal_running_cutoff = state.get("normal_running_cutoff", False)
-            filter_time = self._get_filter_time(custom_config, normal_running_cutoff=normal_running_cutoff)
+            normal_running_cutoff = state.get(self.NORMAL_RUNNING_CUTOFF, False)
+
+            last_log_line = state.get(self.LAST_LOG_LINE, 0)
+            last_runs_filter_time = state.get(self.LAST_RUNS_FILTER_TIME)
+            previous_file_hash = state.get(self.PREVIOUS_FILE_HASH, "")
+
+            max_events_per_run = custom_config.get("max_events_per_run", MAX_EVENTS_PER_RUN_DEFAULT)
+
+            if last_runs_filter_time:
+                # we need to pin the filter time so the same filtering occurs between runs on the same file
+                filter_time = datetime.fromisoformat(last_runs_filter_time)
+            else:
+                filter_time = self._get_filter_time(custom_config, normal_running_cutoff=normal_running_cutoff)
 
             if not header_next_token:
                 self.logger.info("First run...")
+            elif last_log_line > 0:
+                self.logger.info("Subsequent run, continuing to process the previous file...")
             else:
                 self.logger.info("Subsequent run using header_next_token...")
 
             limit_time = time() + 30
             while time() < limit_time:
                 try:
-                    output, headers, status_code = self.connection.client.get_siem_logs(header_next_token)
+                    output, headers, status_code, file_name_list = self.connection.client.get_siem_logs(
+                        header_next_token
+                    )
                     header_next_token = headers.get(self.HEADER_NEXT_TOKEN, header_next_token)
                     if not output:
                         self.logger.info("No new logs returned from Mimecast")
+                        last_log_line = 0
                         break
                 except ApiClientException as error:
                     self.logger.error(
                         f"An exception has been raised during retrieval of siem logs. Status code: {error.status_code} "
                         f"Error: {error}, returning state={state}, has_more_pages={has_more_pages}"
                     )
                     return [], state, has_more_pages, error.status_code, error
-                output = self._filter_and_sort_recent_events(output, filter_time)
+
+                # check if the hashed file list from the previous run is the same as this run
+                if len(output) > max_events_per_run:
```

Comment:
is this condition right? this is to safe guard the token not always pointing to the same file right? so if we do get different output on a subsequent run, it might be less and not go into this and then not reset the index counter, so we might the condition to run the hash if we have a previous one present and use that to reset?

Reply:
this has been updated